runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/DataflowExecutionStateSampler.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker;
+
+import static org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.HashMap;
+import java.util.IntSummaryStatistics;
+import java.util.Map;
+import java.util.Map.Entry;
+import javax.annotation.Nullable;
+import org.apache.beam.runners.core.metrics.ExecutionStateSampler;
+import org.apache.beam.runners.core.metrics.ExecutionStateTracker;
+import org.apache.beam.runners.dataflow.worker.DataflowExecutionContext.DataflowExecutionStateTracker;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.annotations.VisibleForTesting;
+import org.joda.time.DateTimeUtils.MillisProvider;
+
+public class DataflowExecutionStateSampler extends ExecutionStateSampler {
+
+  private static final MillisProvider SYSTEM_MILLIS_PROVIDER = System::currentTimeMillis;
+  private static final DataflowExecutionStateSampler INSTANCE =
+      new DataflowExecutionStateSampler(SYSTEM_MILLIS_PROVIDER);
+
+  private HashMap<String, DataflowExecutionStateTracker> activeTrackersByWorkId = new HashMap<>();
+  private HashMap<String, Map<String, IntSummaryStatistics>> completedProcessingMetrics =
+      new HashMap<>();
+
+  public static DataflowExecutionStateSampler instance() {
+    return INSTANCE;
+  }
+
+  @VisibleForTesting
+  public static DataflowExecutionStateSampler newForTest(MillisProvider clock) {
+    return new DataflowExecutionStateSampler(checkNotNull(clock));
+  }
+
+  public DataflowExecutionStateSampler(MillisProvider clock) {
+    super(clock);
+  }
+
+  @Override
+  public void addTracker(ExecutionStateTracker tracker) {
+    if (!(tracker instanceof DataflowExecutionStateTracker)) {
+      return;
+    }
+    DataflowExecutionStateTracker dfTracker = (DataflowExecutionStateTracker) tracker;
+    this.activeTrackersByWorkId.put(dfTracker.getWorkItemId(), dfTracker);
+  }
+
+  private Map<String, IntSummaryStatistics> mergeStepStatsMaps(
```

Comment:
nit: can this be static? (not sure if it references any member variables or methods)

Reply:
ba58e408fe45269792c997aef6ac0753f9750e8e